### ISSUE TITLE
Cannot read properties of undefined (reading 'isServer')

### ISSUE THREAD

**Issue by fragglebob** [cypress-io/cypress#22291]
### Current behavior

Since upgrading to Cypress 10, we are getting crashes in our CI with the following error:

```
Cannot read properties of undefined (reading 'isServer')
TypeError: Cannot read properties of undefined (reading 'isServer')
    at TLSWrap.onerror (node:_tls_wrap:411:27)
  cypress:network:agent addRequest called { isHttps: true, href: 'https://api.cypress.io/exceptions' } +49ms
  cypress:network:agent got family { family: 4, href: 'https://api.cypress.io/exceptions' } +0ms
  cypress:server:browsers killing browser process +9s
  cypress:server:browsers:chrome closing remote interface client +9s
  cypress:server:browsers:chrome closing chrome +0ms
  cypress:cli child event fired { event: 'exit', code: 1, signal: null } +2m
  cypress:cli Stopping Xvfb +2m
```

This is happening almost ever run we do. Hover we are running in 30 machines parallel, and the most I've seen is two machines fail in the same test run. 

I found an issue https://github.com/nodejs/node/issues/41501 and PR https://github.com/nodejs/node/pull/41523 that relate to this error above, which seems to have been fixed in Node v16.15.0. 

I set up for our tests to run in a custom container image today that is using Node v16.15.1, so I could see if that helps. However I've now realised that cypress has an internal version of node which is being used, v16.13.2 which still has this bug in it from Node. 

### Desired behavior

Ideal if should not being crashing when this error happens.

As this is an issue with how node is happening an error internally, there is probably some other error happening, but it's breaking before we can see what that is. 

### Test code to reproduce

Not sure how to reproduce this as I can't work out what is causing the issue.

In https://github.com/nodejs/node/issues/41501 they weren't able to workout what the exact cause was, other than something wrong with the network somewhere.

### Cypress Version

10.1.0

**Comment by fragglebob** [comment]
Unfortunately I've not seen it locally or anyone else. 

I had already tried that suggestion of a custom images with v16.15.1 and the same browsers.

I think the version being used by cypress's server is the version bundled in electron. I can see this from a couple of logs that report the node version. However it does start sub processes on the system version.

Is there a way to skip the whole electron start up? 

I see electron v20 will have node v16.15.0, but it's still in alpha. 

We have some verbose logging on, and I've been trying to see if there is some pattern to when this happens, but not found anything yet.

**Comment by darrinmn9** [comment]
I am having a similar issue. For me, It occurs when upgrading from `9.6.1` -> `9.7.0 (electron 100)`. For my test suite, I can consistently reproduce this issue locally on my mac and in aws codebuild CI docker image - both of which use node `v16.14.0`. I am not parallelizing tests. What is interesting is that it always fails on the 2nd test in my test suite (the first test passes successfully), but it isn't specific to the actual 2nd test spec. I can move test specs around, comment/delete specs, and it still consistently fails on the 2nd spec that runs.

```
TypeError: Cannot read properties of undefined (reading 'isServer')
    at TLSWrap.onerror (node:_tls_wrap:411:27)
```

**Comment by devhid** [comment]
I'm having the same issue.

This is a (fairly) minimal reproducible example: https://github.com/devhid/cypress-v10-issue

In my case, I'm trying to run Lighthouse on some sample websites for both mobile and desktop viewports, but I get the mentioned error at the end of the test run. 

I do not get this error on older version of Cypress:
```
❯ npx cypress --version    
Cypress package version: 9.6.0
Cypress binary version: 9.6.0
Electron version: 15.5.1
Bundled Node version: 16.5.0
```

Node / NPM versions:
```
❯ node --version  
v16.14.2

❯ npm --version   
8.10.0
```

Cypress Version:
```
❯ npx cypress --version
Cypress package version: 10.1.0
Cypress binary version: 10.1.0
Electron version: 18.0.4
Bundled Node version: 16.13.2
```

Steps:
1. `npm i`
2. `npx cypress open`
3. Run `lighthouse.cy.js`

**Comment by fragglebob** [comment]
I have been able to reproduce the error in the repo @devhid has shared locally. It's doesn't happen every time for me, but it happens often enough.

**Comment by fragglebob** [comment]
Ok, I've done some more research into this error.

It seems the `owner` of `TLSWrap` was being set to something that wasn't a `TLSSocket` by a change in Node v16.7.0 https://github.com/nodejs/node/pull/38468 which was later reverted in version v16.14.0. 

This is happening because the http/s Agent is reusing an existing socket, and this internal change the owner, so when an error does happen it breaks. The only way I can think to avoid a state in these node versions is to not use `keepAlive` and avoid connection reuse. Which would probably degrade performance.

**Comment by flotwig** [comment]
@fragglebob I'm not surprised to find out it's "something wacky to do with Keep-Alive", the interplay of Keep-Alive and Agents is a really tricky area. 

Sounds like the solution is to bump Cypress's internal Node.js version. Looks like [Electron 19.0.0](https://www.electronjs.org/releases/stable#19.0.0) has Node 16.14.2, so it should avoid this.

Would still be nice to have a cleaner reproduction, but at least we have @devhid's sorta flaky repro to test if Electron 19 seems to resolve the issue.

> The only way I can think to avoid a state in these node versions is to not use keepAlive and avoid connection reuse. Which would probably degrade performance.

Yeah, removing Keep-Alive would really hinder proxy performance.

**Comment by fragglebob** [comment]
@flotwig yep, all that async stuff is complicated. 

Yep, a bump to Electron 19 should be good. Which is nicer than having to wait for version 20.

I kinda came up a [little script](https://gist.github.com/fragglebob/ecc6c9db153760128abca04917dd28e3) based on some unit tests in node, where I could quickly see what was going on in the `owner` in different versions.

**Comment by Amirault** [comment]
I have the same behaviour locally and on ci : cypress version : "9.7.0"

**Comment by gavinhenderson** [comment]
I have the same issue on 10.1.0

Is there anything I can do to fix this or do we just have to wait?

**Comment by natapg** [comment]
same issue here after migrating to v10
```
 Cypress:        10.0.3
 Browser:        Custom Chromium 90 (headless)
 Node Version:   v12.22.12
```

**Comment by HannersRipley** [comment]
We had to downgrade to cypress 9.6.1 (which also caused a chain of other downgrades which is very inconvenient). Hopefully this isn't too long in getting fixed or we will need a more involved workaround to prevent this in our CI.

**Comment by fredysiswanto** [comment]
same issue here, run local will force-close
![Screenshot 2022-06-18 092934](https://user-images.githubusercontent.com/45227018/174419136-adc55952-085e-43de-8c8b-c3d17886b610.png)

**Comment by urbany** [comment]
Hi everyone, anyway to circumvent this instead of downgrading to  cypress to 9.6.1 ?
Thanks in advance

**Comment by AtofStryker** [comment]
we are releasing `10.2.0` today, which bumps the electron version to `18.3.0`. Unfortunately, the internal node version I believe is still fixed to [`v16.13.2`](https://github.com/electron/electron/blob/v18.3.0/DEPS#L20), so I do not expect this release to fix this issue. This issue likely won't be resolved until we do a major electron version bump.

**Comment by AtofStryker** [comment]
We are discussing as a team to figure out prioritizing an Electron `19.x` upgrade to resolve issues like this, as well as one thats are related.

**Comment by antoniocl11-w** [comment]
Has there been any solution for this error? The same thing happens to me in github actions

**Comment by aangelinsf** [comment]
Just experienced this with 10.3 on macOS 12.4.

**Comment by Akash141991** [comment]
Facing the same issue on:
- Windows 11
- Cypress 10.3.1
- Chrome 99 (headless) 
- Node v16.14.0
Any workaround?

**Comment by traxtor182** [comment]
Just experienced this with 10.3

**Comment by abinaya-konda** [comment]
experiencing with cypress 10.3, any workaround? This breaks github pipeline!

**Comment by charkour** [comment]
> experiencing with cypress 10.3, any workaround? This breaks github pipeline!

Our workaround was the re-run the workflow a couple of times and then it worked. We are on Node v16 and npm v8 if for some reason those affect Cypress' ability to run.

**Comment by abinaya-konda** [comment]
We are using node 16 and npm 8, running only a few scripts seems to work, but if I execute the entire suite, it always breaks. Is it suggestable to downgrade the cypress version?

**Comment by fragglebob** [comment]
I rolled back to version 9 pretty much after constantly running into this on every suite run. 

Until cypress is based a newer version of electron. This is just gonna keep happening.

**Comment by charkour** [comment]
That's interesting to hear. My team ran into the reported issue once, but haven't had any issues since while using the latest version. 

Not sure why there is a discrepancy between users.

**Comment by fragglebob** [comment]
I guess it comes down to this being caused by network level errors and the way they are handled. 

There must be something in my setup that errors often enough to crash each test suite run.

**Comment by fragglebob** [comment]
Good news. Upgraded today to 10.4.0. All seems resolved so far. 10.4 included the update to electron, which doesn't include a version of node where this is an issue anymore.

**Comment by AtofStryker** [comment]
@fragglebob awesome! Can any others also confirm that `10.4.0` also helps resolve this issue for them?

**Comment by Akash141991** [comment]
Resolved for me after upgrading to 10.4.0 :)

**Comment by AtofStryker** [comment]
I am going to close this since it should be resolved with the electron bump in #22775 which was released in `10.4.0`. If this continues to be an issue for some, please let us know so we can discuss and possibly reopen the issue.

**Comment by HugoBicudo** [comment]
> mesmo problema aqui, executar local irá forçar o fechamento do Windows 11 cypress 10.1.0 navegador Chrome v102 nó v16.15.1
> 
> ![Captura de tela 2022-06-18 092934](https://user-images.githubusercontent.com/45227018/174419136-adc55952-085e-43de-8c8b-c3d17886b610.png)

conseguiu resolver o problema?

**Comment by AtofStryker** [comment]
@HugoBicudo el problema debería resolver en la versión más mejor o igual que `10.4.0`

**Comment by bhadmus** [comment]
This issue has resurfaced in version 12. I am still investigating
